pulp_smash/pulp3/pytest_plugin/__init__.py

```diff
@@ -0,0 +1,236 @@
+    thread: Any
+    ssl_ctx: Any
+    requests_record: Any
+
+    def make_url(self, path):
+        if path[0] != "/":
+            raise ValueError("The `path` argument should start with a '/'")
+
+        if self.ssl_ctx is None:
+            protocol_handler = "http://"
+        else:
+            protocol_handler = "https://"
+
+        return f"{protocol_handler}{self.host}:{self.port}{path}"
+
+
+## Webserver Fixtures
+
+
+@pytest.fixture
+def unused_port():
+    def _unused_port():
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("127.0.0.1", 0))
+            return s.getsockname()[1]
+
+    return _unused_port
+
+
+@pytest.fixture
+def gen_threaded_aiohttp_server(unused_port):
+    fixture_servers_data = []
+
+    def _gen_threaded_aiohttp_server(app, ssl_ctx, call_record):
+        host = cfg.aiohttp_fixtures_origin
+        port = unused_port()
+        shutdown_event = threading.Event()
+        fixture_server = ThreadedAiohttpServer(shutdown_event, app, host, port, ssl_ctx)
+        fixture_server.daemon = True
+        fixture_server.start()
+        fixture_server_data = ThreadedAiohttpServerData(
+            host=host,
+            port=port,
+            shutdown_event=shutdown_event,
+            thread=fixture_server,
+            requests_record=call_record,
+            ssl_ctx=ssl_ctx,
+        )
+        fixture_servers_data.append(fixture_server_data)
+        return fixture_server_data
+
+    yield _gen_threaded_aiohttp_server
+
+    for fixture_server_data in fixture_servers_data:
+        fixture_server_data.shutdown_event.set()
+
+    for fixture_server_data in fixture_servers_data:
+        fixture_server_data.thread.join()
+
+
+@pytest.fixture
+def gen_fixture_server(gen_threaded_aiohttp_server):
+    def _gen_fixture_server(fixtures_root, ssl_ctx):
+        app = web.Application()
+        call_record = add_recording_route(app, fixtures_root)
+        return gen_threaded_aiohttp_server(app, ssl_ctx, call_record)
+
+    yield _gen_fixture_server
+
+
+@pytest.fixture(scope="session")
+def pulpcore_client():
+    configuration = cfg.get_bindings_config()
+    return ApiClient(configuration)
+
+
+@pytest.fixture(scope="session")
+def tasks_api_client(pulpcore_client):
+    return TasksApi(pulpcore_client)
+
+
+## Oprhan Handling Fixtures
+
+
+@pytest.fixture
+def delete_orphans_pre():
+    delete_orphans()
+    yield
+
+
+## Server Side TLS Fixtures
+
+
+@pytest.fixture
+def tls_certificate_authority():
+    return trustme.CA()
+
+
+@pytest.fixture
+def tls_certificate_authority_cert(tls_certificate_authority):
+    return tls_certificate_authority.cert_pem.bytes().decode()
+
+
+@pytest.fixture
+def tls_certificate(tls_certificate_authority):
+    return tls_certificate_authority.issue_cert(
+        cfg.aiohttp_fixtures_origin,
+    )
+
+
+## Client Side TLS Fixtures
+
+
+@pytest.fixture
+def client_tls_certificate_authority():
+    return trustme.CA()
```

Comment:
Is this creating a new CA?
If yes, can we scope it to session?
If not, how is this different from tls_certificate_authority?

Reply:
It is creating a CA with trustme. The tls certificate is the certificate produced with the CA. We can make this session scoped I think.